datalad/cmd.py

```diff
@@ -6,66 +6,86 @@
 #   copyright and license terms.
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command calls, allowing for dry runs and output handling
+
+"""
 
-import os
-import commands
 
+import subprocess
+import sys
 import logging
-lgr = logging.getLogger('datalad.files')
+import os
+
+lgr = logging.getLogger('datalad.cmd')
 
-def dry(s, dry_):
-    """Small helper for dry runs
-    """
-    if dry_:
-        return "DRY " + s
-    return s
 
 class Runner(object):
-    """Helper to run commands and functions while allowing for dry_run and later reporting
+    """Provides a wrapper for system calls.
+
+    An object of this class provides a method to make system calls.
+    Outputs (stdout and stderr) are streamed to stdout during execution as if you were calling it from command line.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `commands`.
     """
 
     __slots__ = ['commands', 'dry']
 
     def __init__(self, dry=False):
+        self.dry = dry
         self.commands = []
-        self.dry = dry                    # TODO: make use of it
 
-    # TODO -- remove dry_run specification here -- use constructor parameter
-    #def __call__(self, cmd, dry_run=False):
-    def getstatusoutput(self, cmd, dry_run=None):
-        """A wrapper around commands.getstatusoutput
+    def run(self, cmd):
+        """Runs the command `cmd` using shell.
+
+        `cmd` is called and uses stdout whereas stderr is captured and logged.
+        In case of dry-mode `cmd` is just added to `commands`.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
         """
-        self.log("Running: %s" % (cmd,))
+
+        self.log("Running: %s" % cmd)
+
         if not self.dry:
-            # status, output = commands.getstatusoutput(cmd)
-            # doing manually for improved debugging
-            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
-            output = ''
-            for line in iter(pipe.readline, ''):
-                self.log("| " + line.rstrip('\n'), level=5)
-                output += line
-            if output[-1:] == '\n': output = output[:-1]
-            status = pipe.close() or 0
+
+            proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=subprocess.PIPE, shell=True)
```

Comment:
ok, as a compromise, just add an option either stdout should be swallowed (logged) or not.  Now I hate how 
```
$> nosetests -s -v --failed datalad/tests/test_main.py                     
#8 datalad.tests.test_main.test_page2annex_same_incoming_and_public('download',) ... 1/
1/1 f.txt
1/d/
1/d/1d
Initialized empty Git repository in /tmp/tmp36D4TQ/.git/
init  ok
(recording state in git...)
add files/1/1 f.txt ok
add files/1/d/1d ok
...
```
looks like.  and then if you like just let stdout be freely spit out where only needed (e.g. during 'get' command) and otherwise just logged